File: src/lucy_orchestrator/state_machine.py
from __future__ import annotations

from .exceptions import InvalidTransitionError
from .models import Task, TaskState


_ALLOWED_TRANSITIONS: dict[TaskState, set[TaskState]] = {
    TaskState.NEW: {TaskState.CLARIFYING, TaskState.FAILED, TaskState.CANCELLED},
    TaskState.CLARIFYING: {
        TaskState.WAIT_APPROVAL,
        TaskState.FAILED,
        TaskState.CANCELLED,
    },
    TaskState.WAIT_APPROVAL: {TaskState.RUNNING, TaskState.FAILED, TaskState.CANCELLED},
    TaskState.RUNNING: {TaskState.TESTING, TaskState.FAILED, TaskState.CANCELLED},
    TaskState.TESTING: {TaskState.DONE, TaskState.FAILED, TaskState.CANCELLED},
    TaskState.FAILED: {TaskState.RUNNING, TaskState.CANCELLED},
    TaskState.DONE: set(),
    TaskState.CANCELLED: set(),
}


def assert_transition(task: Task, target: TaskState) -> None:
    allowed = _ALLOWED_TRANSITIONS.get(task.state, set())
    if target not in allowed:
        raise InvalidTransitionError(
            f"Invalid transition: {task.state.value} -> {target.value}"
        )

    if target == TaskState.RUNNING:
        if task.approval.required and not task.approval.is_approved():
            raise InvalidTransitionError("Task approval is required before RUNNING")
        if task.plan is None:
            raise InvalidTransitionError("Task plan is required before RUNNING")
        open_questions = task.plan.open_required_questions()
        if open_questions:
            ids = ", ".join(item.id for item in open_questions)
            raise InvalidTransitionError(f"Open required questions remain: {ids}")

    if target == TaskState.TESTING and not task.artifacts.diff_path:
        raise InvalidTransitionError("Diff artifact is required before TESTING")

    if target == TaskState.DONE and not task.artifacts.test_report_path:
        raise InvalidTransitionError("Test report is required before DONE")


def transition(
    task: Task,
    target: TaskState,
    event_type: str,
    message: str,
    payload: dict[str, object] | None = None,
) -> None:
    assert_transition(task, target)
    previous = task.state
    task.state = target
    event_payload = {"from": previous.value, "to": target.value}
    if payload:
        event_payload.update(payload)
    task.record_event(event_type=event_type, message=message, payload=event_payload)

File: src/state-machine.ts
import { InvalidTransitionError } from "./errors.js"
import { isApproved, openRequiredQuestions, recordTaskEvent, TaskState, type Task } from "./models.js"

const ALLOWED_TRANSITIONS: Record<TaskState, Set<TaskState>> = {
  [TaskState.NEW]: new Set([TaskState.CLARIFYING, TaskState.FAILED, TaskState.CANCELLED]),
  [TaskState.CLARIFYING]: new Set([TaskState.WAIT_APPROVAL, TaskState.FAILED, TaskState.CANCELLED]),
  [TaskState.WAIT_APPROVAL]: new Set([TaskState.RUNNING, TaskState.FAILED, TaskState.CANCELLED]),
  [TaskState.RUNNING]: new Set([TaskState.TESTING, TaskState.FAILED, TaskState.CANCELLED]),
  [TaskState.TESTING]: new Set([TaskState.DONE, TaskState.FAILED, TaskState.AUTO_FIXING, TaskState.CANCELLED]),
  [TaskState.AUTO_FIXING]: new Set([TaskState.TESTING, TaskState.DONE, TaskState.FAILED, TaskState.CANCELLED]),
  [TaskState.DONE]: new Set(),
  [TaskState.FAILED]: new Set([TaskState.RUNNING, TaskState.AUTO_FIXING, TaskState.CANCELLED]),
  [TaskState.CANCELLED]: new Set(),
}

export function assertTransition(task: Task, target: TaskState): void {
  const allowed = ALLOWED_TRANSITIONS[task.state]
  if (!allowed?.has(target)) {
    throw new InvalidTransitionError(`Invalid transition: ${task.state} -> ${target}`)
  }

  if (target === TaskState.RUNNING) {
    if (!isApproved(task)) {
      throw new InvalidTransitionError("Task approval is required before RUNNING")
    }
    if (!task.plan) {
      throw new InvalidTransitionError("Task plan is required before RUNNING")
    }
    const openQuestions = openRequiredQuestions(task)
    if (openQuestions.length > 0) {
      throw new InvalidTransitionError(
        `Open required questions remain: ${openQuestions.map((item) => item.id).join(", ")}`,
      )
    }
  }

  if (target === TaskState.TESTING && !task.artifacts.diffPath) {
    throw new InvalidTransitionError("Diff artifact is required before TESTING")
  }

  if (target === TaskState.DONE && !task.artifacts.testReportPath) {
    throw new InvalidTransitionError("Test report is required before DONE")
  }
}

export function transition(
  task: Task,
  target: TaskState,
  eventType: string,
  message: string,
  payload: Record<string, unknown> = {},
): void {
  assertTransition(task, target)
  const previous = task.state
  task.state = target

  recordTaskEvent(task, eventType, message, {
    from: previous,
    to: target,
    ...payload,
  })
}
